funlet-forward/functions/funlet-forward.js

```diff
@@ -0,0 +1,419 @@
+    with given caller ID and timeout, and to redirect to the given
+    fallback URL if the forwarding call fails.
+    When the caller is a restricted number, the response is modified
+    to play or say an error message instead.
+*/
+function forwardStage1(
+  response,
+  isForwardingAllowed, accessRestrictedErrorMessage, language, voice,
+  callerId, forwardingNumber, timeout, fallbackUrl
+) {
+  if ( !isForwardingAllowed ) {
+    simpleMessage(response, accessRestrictedErrorMessage, language, voice)
+    return;
+  }
+  let dialOptions = {
+    action: getForwardActionUrl( fallbackUrl ),
+  };
+  if ( callerId !== "" ) {
+    dialOptions.callerId = callerId;
+  }
+  dialOptions.timeout = timeout;
+  response.dial( dialOptions, forwardingNumber );
+}
+
+/*
+  Function: forwardStage2()
+
+  Parameters:
+    * response - Twilio.twiml.VoiceResponse, Twilio Voice response in progress
+    * isDialDone - boolean, whether the forwarding call has completed
+    * callStatus - string, the status of the forwarding call
+    * fallbackUrl - string, URL of a script with further instructions
+                    in case the forwarding call failed
+  Response:
+    Until the forwarding call has completed, the response is left unchanged.
+    When the forwarding call has ended in a failure, the response is modified
+    with instructions to redirect to the fallback URL, if any. Otherwise,
+    an instruction to hang up is added to the response.
+
+  Returns:
+    true when the forwarding call has completed,
+    false otherwise.
+*/
+function forwardStage2(response, isDialDone, callStatus, fallbackUrl) {
+  if (isDialDone) {
+    if (
+      callStatus !== "answered" &&
+      callStatus !== "completed" &&
+      fallbackUrl !== ""
+    ) {
+      response.redirect( fallbackUrl );
+    } else {
+      response.hangup();
+    }
+  }
+  return isDialDone;
+}
+
+/*
+  4. Main Handler
+
+  This is the entry point to your Twilio Function,
+  which will run to process an incoming HTTP request
+  such as the ones generated by Twilio events.
+*/
+
+exports.handler = function(context, event, callback) {
+  const NO_ERROR = null;
+
+  let
+    response = new Twilio.twiml.VoiceResponse(),
+    isDial = isDialDone(event, context, config),
+    callStatus = getCallStatus(event, context, config),
+    fallbackUrl = getFallbackUrl(event, context, config),
+    caller = getCaller(event, context, config),
+    called = getPhoneNumberCalled(event, context, config),
+    allowedCallers = getAllowedCallers(event, context, config),
+    accessRestrictedErrorMessage =
+      getAccessRestrictedErrorMessage(event, context, config),
+    language = getLanguage(event, context, config),
+    voice = getVoice(event, context, config),
+    callerId = getCallerId(event, context, config),
+    forwardingNumber = getPhoneNumber(event, context, config),
+    timeout = getTimeout(event, context, config);
+
+  if (
+    !forwardStage2( response, isDial, callStatus, fallbackUrl )
```

Comment:
You are right: the order of the calls does not reflect the order of the stages. They are called by order of specificity: stage 2 is triggered by the presence of the `Dial` flag, otherwise we are at stage 1.

I don't mind renaming these functions (it's the second time it comes up!) and I am open to suggestions.

The current names are based on:
* the name of the Twimlet
* the stage of its processing:
  - stage 1 corresponds to the entry point
  - stage 2 is the reentry for the first callback
  - and so on

Reply:
I think for this one we should do two things. 
1) reorder the logic it makes it a bit longer but more readable in my opinion
2) rename the functions but specify the stages in comments.

Basically something like this:

```js
if (!isDialDone) {
	// stage 1
	// check if call is valid 
	forwardCallOrPlayMessage(
      response,
      isForwardingAllowed(caller, called, allowedCallers),
      accessRestrictedErrorMessage, language, voice,
      callerId, forwardingNumber, timeout, fallbackUrl
    );
} else {
	// stage 2
	// forwarded call has been finished
	handleFinishedCall(response, isDialDone, callStatus, fallbackUrl);
}
```

That way you can read it a bit more in the execution in which the stages are actually being run. Which hopefully helps people read the code a bit better.

Reply:
This is not what I had in mind. The idea was to keep the routing logic out of the main handler, which is why I let the function for stage 2 check whether it was its turn to run, and return the result of this check as a boolean.

I would rather remove more logic from the main handler, and if the routing is not done in each stage, to avoid running them in a non-intuitive order, it should be delegated to a separate routing function.

If I go further this way, I would also remove the reading of input parameters from the main handler, and make each stage function a handler with the same signature. Without presuming of the proper names for these handlers, here called `handler1` and `handler2` temporarily for the duration of the argument, this would give something along these lines:

```
let response = null;
function getResponse() {
  if ( response === null ) {
    response = new Twilio.twiml.VoiceResponse();
  }
  return response;
}

function handler1(context, event, callback) {
  let
    response = getResponse(),
    caller = getCaller(event, context, config),
    called = getPhoneNumberCalled(event, context, config),
    allowedCallers = getAllowedCallers(event, context, config),
    accessRestrictedErrorMessage = getAccessRestrictedErrorMessage(event, context, config),
    language = getLanguage(event, context, config),
    voice = getVoice(event, context, config),
    callerId = getCallerId(event, context, config),
    forwardingNumber = getPhoneNumber(event, context, config),
    timeout = getTimeout(event, context, config);

  if ( !isForwardingAllowed(caller, called, allowedCallers) ) {
    simpleMessage(response, accessRestrictedErrorMessage, language, voice)
    return;
  }
  let dialOptions = {
    action: getForwardActionUrl( fallbackUrl ),
  };
  if ( callerId !== "" ) {
    dialOptions.callerId = callerId;
  }
  dialOptions.timeout = timeout;
  response.dial( dialOptions, forwardingNumber );
}

function handler2(context, event, callback) {
  let
    response = getResponse(),
    callStatus = getCallStatus(event, context, config),
    fallbackUrl = getFallbackUrl(event, context, config);

  if ( callStatus !== "answered" &&  callStatus !== "completed" &&  fallbackUrl !== "" ) {  
    response.redirect( fallbackUrl );  
  } else {  
    response.hangup();  
  }
}

function getStage(context, event) {
  if isDialDone(event, context, config) {
    return "stage2";
  } else {
    return "stage1";
  }
}

exports.handler = function(context, event, callback) {
  let
    routes = {"stage1": handler1, "stage2": handler2},
    stage = getStage(context, event);

  routes[stage](context, event, callback);
}
```